### ISSUE TITLE
"Failed to obtain locks until deadline" message repeatedly appearing in the log

### ISSUE THREAD

**Issue by mbautin** [yugabyte/yugabyte-db#4269]
This is affecting write traffic on this test cluster. Reported by @rao-vasireddy.

![image](https://user-images.githubusercontent.com/552936/79931559-1f680a80-8400-11ea-8775-26d0be920591.png)

**Comment by mbautin** [comment]
Attaching a log from one of the nodes obtained using this command:
```
( zgrep "locks until" *.gz; grep -H "locks until" *.INFO ) | sort -t: -k2,2 >lock_timeout_grep_sorted_$( date +%Y-%m-%dT%H_%M_%S ).tx
```
[n1_lock_timeout_grep_sorted_2020-04-22T01_20_31.txt.gz](https://github.com/yugabyte/yugabyte-db/files/4513399/n1_lock_timeout_grep_sorted_2020-04-22T01_20_31.txt.gz)

**Comment by mbautin** [comment]
The same for the third node:
[n3_lock_timeout_grep_sorted_2020-04-22T01_20_25.txt.gz](https://github.com/yugabyte/yugabyte-db/files/4513410/n3_lock_timeout_grep_sorted_2020-04-22T01_20_25.txt.gz)

**Comment by mbautin** [comment]
And for the second node (4 parts as separate files):
[n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part1.txt.gz](https://github.com/yugabyte/yugabyte-db/files/4513426/n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part1.txt.gz)
[n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part4.txt.gz](https://github.com/yugabyte/yugabyte-db/files/4513427/n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part4.txt.gz)
[n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part3.txt.gz](https://github.com/yugabyte/yugabyte-db/files/4513428/n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part3.txt.gz)
[n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part2.txt.gz](https://github.com/yugabyte/yugabyte-db/files/4513429/n2_lock_timeout_grep_sorted_2020-04-22T01_20_14_part2.txt.gz)

**Comment by mbautin** [comment]
Currently running YSQL statements:
- Node 1: https://gist.githubusercontent.com/mbautin/f648e083a677cef358a7518d7a2c8846/raw
- Node 2: https://gist.github.com/mbautin/cf1af627ddd480ae683c42273b63e6f4/raw
- Node 3: https://gist.githubusercontent.com/mbautin/5914c6c8ba4c19b67cd9c6d0ee83e45d/raw

**Comment by mbautin** [comment]
It looks like the issue started happening around 5:28 PM UTC (10:28 AM PDT) on April 18th, 2020. This matches both the drop in IOPS in the graph below and the start of log spew in the attached files.

![image](https://user-images.githubusercontent.com/552936/79932939-7b805e00-8403-11ea-9cfe-8510680cf0b7.png)

**Comment by bmatican** [comment]
Some more notes, went to N1, can see very long running (stuck?) YCQL ops:
```
                {
                    "elapsed_millis": 119244,
                    "cql_details": {
                        "type": "EXECUTE",
                        "call_details": [
                            {
                                "sql_id": "2deb94664b0596029e02ff391e4a8ca6",
                                "sql_string": "INSERT INTO CassandraSecondaryIndex (k, v) VALUES (?, ?);"
                            }
                        ]
                    }
                },
```

**Comment by bmatican** [comment]
Seems like local writes are making progress fine, but the remote ones are failing (60s timeout)?

<img width="448" alt="Screen Shot 2020-04-22 at 10 30 56 AM" src="https://user-images.githubusercontent.com/395011/80013885-6ef01a00-8484-11ea-8bcc-be631406c8c6.png">

**Comment by bmatican** [comment]
Seems like N2 has a lot of `rpc_tp_TabletServer` threads stuck in `LockedBatchEntry::Lock()`. Refreshing the page after 3s seems to still show many of the same thread IDs, which suggests they're actually not timing out after the deadlines, as expected...

cc @spolitov @mbautin

**Comment by bmatican** [comment]
@spolitov pointed out that N2 seems stuck on
```
    @     0x7f940ff9511f  (unknown)
    @     0x7f941091346c  __GI___nanosleep
    @     0x7f941e225989  yb::tablet::SnapshotOperation::AcquireSchemaLock()
    @     0x7f941e19707f  yb::tablet::TabletSnapshots::Prepare()
    @     0x7f941e224ae3  yb::tablet::SnapshotOperation::Prepare()
    @     0x7f941e21ed8e  yb::tablet::OperationDriver::PrepareAndStart()
    @     0x7f941e24d02b  yb::tablet::PreparerImpl::ProcessAndClearLeaderSideBatch()
    @     0x7f941e24d43f  yb::tablet::PreparerImpl::Run()
    @     0x7f94160d81f3  yb::ThreadPool::DispatchThread()
    @     0x7f94160d48fe  yb::Thread::SuperviseThread()
    @     0x7f941090b693  start_thread
    @     0x7f941004841c  __clone
    @ 0xffffffffffffffff  (unknown)
```

However, there seems to be no other thread doing DDL.

Checking the master `/tasks` endpoint, we see  there are active snapshot tasks in progress, that keep getting retried:
```
Task Name | State | Start Time | Time | Description
-- | -- | -- | -- | --
kWaiting | 12.2 min ago | 12.2 min | 4a901240b5fd432db5acfdfbcdce763f (table cassandrasecondaryindexbyvalue [id=e24444b6e25a4a258bfdcb3365142956]) Tablet Snapshot Operation 3 RPC
kWaiting | 12.3 min ago | 12.3 min | 4a901240b5fd432db5acfdfbcdce763f (table cassandrasecondaryindexbyvalue [id=e24444b6e25a4a258bfdcb3365142956]) Tablet Snapshot Operation 3 RPC
```

This is happening just for this tablet `4a901240b5fd432db5acfdfbcdce763f`, but over and over again.

Given that, my suspicion was that the only way we could jam the Prepare thread, was if it is the source of the deadlock itself...

Doing some log zgrep on N2, checking for all tablets
`zgrep 'Processing TabletSnapshotOp for tablet' *INFO* > all_snapshots.log`

We see a bunch of retries for all tablets, until finally just the one that's now stuck remains. Tracing back when it started we see:
```
I0418 17:25:13.597187 22413 backup_service.cc:60] Processing TabletSnapshotOp for tablet 4a901240b5fd432db5acfdfbcdce763f from 172.151.27.186:48455
I0418 17:25:13.597229 22413 backup_service.cc:60] Processing TabletSnapshotOp for tablet 4a901240b5fd432db5acfdfbcdce763f from 172.151.27.186:45131
```

At this point, I'm assuming the Prepare thread picked up both of these ops and tried to prepare them, but in Prepare we take the schema lock, so the thread tried to lock the same schema twice and is now deadlocked.

cc @spolitov @mbautin if that makes sense

**Comment by bmatican** [comment]
Oh, and timing wise, the 2 quick ops line up with the rolling restart for software upgrade:
```
UpgradeSoftware Universe | puppy-food-5 | Completed | April 18, 2020, 10:21 AM | April 18, 2020, 10:33 AM
```

Checking this master leader logs:
> I0418 17:24:41.902966   405 sys_catalog.cc:366] T 00000000000000000000000000000000 P bf2ce53f64204d5c98db66e65cb0d0ae [sys.catalog]: This master's current role is: LEADER

Corresponding 2 tasks being fired for this tablet:
```
W0418 17:25:13.567745   633 async_rpc_tasks.cc:231] 4a901240b5fd432db5acfdfbcdce763f (table cassandrasecondaryindexbyvalue [id=e24444b6e25a4a258bfdcb3365142956]) Tablet Snapshot Operation 3 RPC (task=0x467a010, state=kRunning): TS 8d4223e90007476c8fd8794a84fd7b0f: Tablet Snapshot Operation RPC failed for tablet 4a901240b5fd432db5acfdfbcdce763f: Timed out (yb/rpc/outbound_call.cc:512): TabletSnapshotOp RPC (request call id 159) to 172.151.31.1:9100 timed out after 30.000s
I0418 17:25:13.567770   633 async_rpc_tasks.cc:303] 4a901240b5fd432db5acfdfbcdce763f (table cassandrasecondaryindexbyvalue [id=e24444b6e25a4a258bfdcb3365142956]) Tablet Snapshot Operation 3 RPC (task=0x467a010, state=kRunning): Scheduling retry with a delay of 29ms (attempt = 1)...
W0418 17:25:13.567791   633 async_rpc_tasks.cc:231] 4a901240b5fd432db5acfdfbcdce763f (table cassandrasecondaryindexbyvalue [id=e24444b6e25a4a258bfdcb3365142956]) Tablet Snapshot Operation 3 RPC (task=0x2ca0eb0, state=kRunning): TS 8d4223e90007476c8fd8794a84fd7b0f: Tablet Snapshot Operation RPC failed for tablet 4a901240b5fd432db5acfdfbcdce763f: Timed out (yb/rpc/outbound_call.cc:512): TabletSnapshotOp RPC (request call id 111) to 172.151.31.1:9100 timed out after 30.000s
I0418 17:25:13.567800   633 async_rpc_tasks.cc:303] 4a901240b5fd432db5acfdfbcdce763f (table cassandrasecondaryindexbyvalue [id=e24444b6e25a4a258bfdcb3365142956]) Tablet Snapshot Operation 3 RPC (task=0x2ca0eb0, state=kRunning): Scheduling retry with a delay of 29ms (attempt = 1)...
```

Why would there even be 2 tasks for this though? @spolitov

**Comment by bmatican** [comment]
So currently I still see a number of open question:
- According to YW, the backup on the 14th completed successfully, but we clearly had a leftover Snapshot operation in flight. **How come we reported success, if we still had ops in flight?**
- When we restarted the cluster on the 18th, **how come the master triggered 2 snapshot operations for some tablets** (`4a901240b5fd432db5acfdfbcdce763f`) included.
- Issuing Alter operations in quick succession (ie: in the same ms), such that they could grab the schema lock in the same batch is obvious deadlock material. However, normally that should not happen. Immediate question maybe is: **do we need the schema lock for snapshots?**
- Even with the schema lock deadlock on prepare, why are cluster writes having issues? I thought we have a Preparer per tablet, but with a shared thread pool, so I can understand writes jamming for the 1 tablet that's stuck in prepare, but why are all writes dropping to 0? Is this workload a batch workload?

**Comment by rao-vasireddy** [comment]
@bmatican , this is not a batch workload

**Comment by mbautin** [comment]
> According to YW, the backup on the 14th completed successfully, but we clearly had a leftover Snapshot operation in flight. How come we reported success, if we still had ops in flight?

Maybe a retry succeeded and one of the original ops got stuck.

**Comment by bmatican** [comment]
Went back to YW to check the order of operations:
```
UpgradeSoftware Universe | puppy-food-5 | Completed | April 18, 2020, 10:21 AM | April 18, 2020, 10:33 AM |  
-- | -- | -- | -- | -- | --
Restore Backup | cassandrasecondaryindex2 | Completed | April 15, 2020, 12:01 AM | April 15, 2020, 12:09 AM |  
Backup Table | sqlsecondaryindex | Failed | April 14, 2020, 11:57 PM | April 14, 2020, 11:59 PM | See Details
Backup Table | cassandrasecondaryindex | Completed | April 14, 2020, 11:53 PM | April 14, 2020, 11:56 PM
```

So backup of `cassandrasecondaryindex` at 11:53, then restore, as different table name? `cassandrasecondaryindex2` at 12:01

Checking that time's master logs.

11:53 backup and our table's snapshot finishing:
```
I0415 06:53:17.533046  3983 catalog_manager_ent.cc:274] Servicing CreateSnapshot request: tables { table_name: "cassandrasecondaryindex" namespace { id: "16a3f89a00e943fdaef7e15958664578" name: "ybdemo_keyspace" database_type: YQL_DATABASE_CQL } } transaction_aware: true add_indexes: true
...
I0415 06:53:17.757845 31480 master_snapshot_coordinator.cc:153] Finished CREATING snapshot at 4a901240b5fd432db5acfdfbcdce763f: 2
```

12:01 restore however also does another snapshot?
```
I0415 07:01:31.816761  8982 catalog_manager_ent.cc:819] Servicing ImportSnapshotMeta request
...
I0415 07:01:31.831068  8982 catalog_manager.cc:1909] CreateTable from internal request:
name: "cassandrasecondaryindex2"
...
I0415 07:01:32.528771  1204 catalog_manager_ent.cc:274] Servicing CreateSnapshot request: tables { table_id: "ad10c050724e429484f8894420782a4b" } tables { table_id: "e24444b6e25a4a258bfdcb3365142956" } transaction_aware: true add_indexes: false
W0415 07:01:32.569614  8982 scoped_leader_shared_lock.cc:114] Long lock of catalog manager: 0.712s
    @     0x7f47a17f9490  yb::master::ScopedLeaderSharedLock::Unlock()
    @     0x7f47a18fd579  yb::master::MasterBackupServiceImpl::ImportSnapshotMeta()
...
I0415 07:01:32.856976  1288 master_snapshot_coordinator.cc:153] Finished CREATING snapshot at 4a901240b5fd432db5acfdfbcdce763f: 2
```

**Comment by mbautin** [comment]
An observation from @spolitov: for operations acquiring the schema lock, we release it when the operation gets replicated, but if it has been aborted, we rely on the destructor to release the schema lock. However, if the aborted operation is referenced from somewhere, the destructor won't be called yet.

Created https://github.com/yugabyte/yugabyte-db/issues/4278 to track this.

**Comment by mbautin** [comment]
`/threadz?group=all` dump (obtained using `lynx --dump --width=200 "<url>"`):
- Node 1: https://gist.githubusercontent.com/mbautin/d8886842cdcd18d8af1eea27abfc37c0/raw
- Node 2: https://gist.githubusercontent.com/mbautin/e056a395845a87b46622d4ac530d7554/raw
- Node 3: https://gist.githubusercontent.com/mbautin/3b708b8df665a9164e1cdacb30abb16c/raw

**Comment by bmatican** [comment]
I was able to get a repro for this:
- create universe and start table+index workload
- backup table from YW
- restore table back
- restart the cluster (I think only the masters are necessary)

After restart, the master seems to re-trigger snapshot operations.

Also, I confirmed it's specific to restore, because initially I had also restarted the cluster right after the backups and nothing odd happened!
cc @mbautin @spolitov

**Comment by spolitov** [comment]
Fixed by a7fd656ca19504d6ca9104ec72f70ecba35086b7
